File: typescript/packages/ampersend-sdk/src/x402/http/factory.ts
/**
 * Returns an `x402Client` subclass, so the result drops into
 * `wrapFetchWithPayment` unchanged. For advanced setups, construct
 * `AmpersendX402Client` directly.
 */

import { EVM_NETWORK_CHAIN_ID_MAP } from "@x402/evm/v1"
import type { Address, Hex } from "viem"

import { createAmpersendTreasurer } from "../../ampersend/treasurer.ts"
import { AmpersendX402Client } from "./client.ts"

const DEFAULT_API_URL = "https://api.ampersend.ai"
const DEFAULT_NETWORK = "base"

export interface SimpleHttpClientOptions {
  /** Smart account address. */
  smartAccountAddress: Address
  /** Session key private key for signing. */
  sessionKeyPrivateKey: Hex
  /** Ampersend API URL. Defaults to production. */
  apiUrl?: string
  /** v1 network name (e.g. `"base"`). v1 and v2 (CAIP-2) are both registered. */
  network?: string
}

/**
 * Create an `AmpersendX402Client` wired to the Ampersend API.
 *
 * @example
 * ```typescript
 * import { wrapFetchWithPayment } from "@x402/fetch"
 * import { createAmpersendHttpClient } from "@ampersend_ai/ampersend-sdk"
 *
 * const client = createAmpersendHttpClient({
 *   smartAccountAddress: "0x...",
 *   sessionKeyPrivateKey: "0x...",
 * })
 * const fetchWithPay = wrapFetchWithPayment(fetch, client)
 * const response = await fetchWithPay("https://paid-api.com/endpoint")
 * ```
 */
export function createAmpersendHttpClient(options: SimpleHttpClientOptions): AmpersendX402Client {
  const network = options.network ?? DEFAULT_NETWORK
  const chainId = (EVM_NETWORK_CHAIN_ID_MAP as Readonly<Record<string, number>>)[network]
  if (chainId === undefined) {
    throw new Error(`Unknown network: ${network}`)
  }

  const treasurer = createAmpersendTreasurer({
    smartAccountAddress: options.smartAccountAddress,
    sessionKeyPrivateKey: options.sessionKeyPrivateKey,
    apiUrl: options.apiUrl ?? DEFAULT_API_URL,
  })

  return new AmpersendX402Client(treasurer).withNetworks({
    v1: [network],
    v2: [`eip155:${chainId}`],
  })
}

File: rust/ampersend-sdk/src/http/factory.rs
use std::sync::Arc;

use crate::ampersend::treasurer::{
    create_ampersend_treasurer, AmpersendTreasurerConfig, SimpleAmpersendTreasurerConfig,
};
use crate::http::adapter::X402HttpClient;
use crate::x402::types::network_to_chain_id;

/// Default Ampersend API URL.
const DEFAULT_API_URL: &str = "https://api.ampersend.ai";

/// Default network (Base mainnet for production).
const DEFAULT_NETWORK: &str = "base";

/// Simplified options for Ampersend HTTP client wrapper.
pub struct SimpleHttpClientOptions {
    /// Smart account address
    pub smart_account_address: String,
    /// Session key private key for signing
    pub session_key_private_key: String,
    /// Ampersend API URL (defaults to production)
    pub api_url: Option<String>,
    /// Network to use (defaults to "base"). Chain ID is inferred from this.
    pub network: Option<String>,
}

/// Create an x402 HTTP client with Ampersend payment support.
///
/// This integrates ampersend-sdk with automatic payment handling for HTTP requests.
pub fn create_ampersend_http_client(
    options: SimpleHttpClientOptions,
) -> Result<X402HttpClient, crate::error::SdkError> {
    let network = options.network.as_deref().unwrap_or(DEFAULT_NETWORK);
    let chain_id = network_to_chain_id(network)
        .ok_or_else(|| crate::error::SdkError::Config(format!("Unknown network: {network}")))?;

    let treasurer = create_ampersend_treasurer(AmpersendTreasurerConfig::Simple(
        SimpleAmpersendTreasurerConfig {
            smart_account_address: options.smart_account_address,
            session_key_private_key: options.session_key_private_key,
            api_url: Some(
                options
                    .api_url
                    .unwrap_or_else(|| DEFAULT_API_URL.to_string()),
            ),
            chain_id: Some(chain_id),
        },
    ));

    Ok(X402HttpClient::new(Arc::new(treasurer)))
}
